Added option to disable Sign Dyeing

werogg opened the pull request:
Fixes #2948

werogg committed:
Add disable sign dyeing option to config

werogg committed:
Add new ProtectConfig to disable sign dyeing

werogg committed:
Add method to cancel sign dyeing event

montlikadani commented on EssentialsProtect/src/com/earth2me/essentials/protect/EssentialsProtectBlockListener.java:
```diff
@@ -90,4 +92,44 @@ public void onPortalLight(PortalCreateEvent event) {
+                        itemInHandMat == Material.PURPLE_DYE ||
+                        itemInHandMat == Material.RED_DYE ||
+                        itemInHandMat == Material.WHITE_DYE ||
+                        itemInHandMat == Material.YELLOW_DYE)) {
```
You should use the [MaterialUtil](https://github.com/EssentialsX/Essentials/blob/2.x/Essentials/src/com/earth2me/essentials/utils/MaterialUtil.java) class, which you can check easily items.

montlikadani commented on Essentials/src/config.yml:
```diff
@@ -844,6 +844,9 @@ protect:
     # Should Wither damage be disabled?
     wither: false
 
+    #Should sign dyeing be disabled?
```
Put space before "#" signal, to the text is be readable.

werogg replied:
Solved

werogg committed:
missing space

triagonal commented on EssentialsProtect/src/com/earth2me/essentials/protect/EssentialsProtectBlockListener.java:
```diff
@@ -90,4 +92,44 @@ public void onPortalLight(PortalCreateEvent event) {
+    public void onSignDye(PlayerInteractEvent event) {
+        Material clickedBlock = event.getClickedBlock().getBlockData().getMaterial();
+        Action action = event.getAction();
+        Material itemInHandMat = event.getPlayer().getInventory().getItemInMainHand().getType();
```
This will error out in 1.8.8, as `getItemInMainHand()` didn't exist until 1.9. If I'm not mistaken, the preferred way to handle this is to get the Essentials `User` and use `User#getItemInHand()` which accounts for both cases.

werogg replied:
Fixed on commit 2e5296881436ed9fe5cb967a39095de32613f261

pop4959 replied:
What also would have probably worked just fine is to exit this method early by checking the MineCraft version (since this is only intended to run on versions 1.15+). See VersionUtil for some handy methods.

Edit: Although I still completely agree with triangle, any code that *can* be written in a version independent way should absolutely be.

werogg committed:
Added isDye() check to MaterialUtil

werogg committed:
Code simplification

werogg committed:
Fix getItemInMainHand() didn't exist until 1.9
Now using User#getItemInHand()

werogg commented:
I ask because I can't test it now, but could be possible to avoid this "sign-dyeing protection" by using the off hand? (Second hand). If it's possible then I will have to consider that option.
Will test it as soon as I can.

EDIT: It is not possible ^^

pop4959 commented:
I would suggest implementing this as a permission rather than a config option because it will make controlling the feature much easier, as well as adding less extra bloat to the config.

Something like `essentials.signs.color.dye` or simply `essentials.signs.dye` may be reasonable, given that the `essentials.signs.color` permission node already exists. Default value of the permission should be `true` though to retain vanilla behaviour.

The most likely use case of this feature is server owners who want to disable sign dying for players who do not have the `essentials.signs.color` permission.

mdcfe commented on Essentials/src/com/earth2me/essentials/utils/MaterialUtil.java:
```diff
@@ -68,6 +69,11 @@
             "ACACIA_WALL_SIGN", "BIRCH_WALL_SIGN", "DARK_OAK_WALL_SIGN", "JUNGLE_WALL_SIGN",
             "OAK_WALL_SIGN", "SPRUCE_WALL_SIGN");
+
+        DYES = EnumUtil.getAllMatching(Material.class, "DYE", "BLACK_DYE", "BLUE_DYE", "BROWN_DYE",
```
This should either explicitly note that these are only the dyes found in 1.14+, or it should include the dyes from before 1.14. I'd recommend the former as dyeing signs is 1.15+.

triagonal replied:
Just a sidenote, but sign dyeing was actually implemented [in 1.14](https://minecraft.gamepedia.com/Sign#History), not 1.15. Doesn't change anything though, just for reference.

werogg replied:
That Set already includes all the dyes before 1.14, but could be a good practice to separe them and note it.

mdcfe commented:
@pop4959 If this PR was for EssentialsX itself, I'd agree, but EssentialsX Protect is a config-controlled plugin and doesn't generally use permissions-based control. This approach is suitable for EssentialsX Protect unless we want to move this feature into EssentialsX itself.

pop4959 commented:
I didn't mention it in the previous message but I was on the fence about whether this really fits in EssentialsXProtect, but upon looking into it further I guess that really is where it would best belong (as a physics/behaviour modification). So I guess it doesn't matter too much.

werogg committed:
MaterialUtil.isDye() is now 1.14+ friendly and split for future uses

Sign dyeing event handler optimized for non-used versions (Before 1.14)

PlayerInteractEvent#getItem instead of using User Object

Convention fix

Co-Authored-By: md678685 <1917406+md678685@users.noreply.github.com>

Config.yml comments fix

Co-Authored-By: md678685 <1917406+md678685@users.noreply.github.com>

v1_14_DYES usages fixed to V1_14_DYES

werogg committed:
MaterialUtil.isDye() is now 1.14+ friendly and split for future uses

werogg committed:
Sign dyeing event handler optimized for non-used versions (Before 1.14)

mdcfe commented on Essentials/src/com/earth2me/essentials/utils/MaterialUtil.java:
```diff
@@ -27,6 +27,8 @@
     private static final Set<Material> POTIONS;
     private static final Set<Material> SIGN_POSTS;
     private static final Set<Material> WALL_SIGNS;
+    private static final Set<Material> v1_14_DYES;
```
```suggestion
    private static final Set<Material> V1_14_DYES;
```

\+ change any usages of this field

mdcfe commented on Essentials/src/config.yml:
```diff
@@ -844,6 +844,9 @@ protect:
     # Should Wither damage be disabled?
     wither: false
 
+    # Should sign dyeing be disabled?
```
```suggestion
    # Should players be able to dye signs?
    # This does not affect the essentials.sign.color permission.
```

mdcfe commented on EssentialsProtect/src/com/earth2me/essentials/protect/EssentialsProtectBlockListener.java:
```diff
@@ -90,4 +97,20 @@ public void onPortalLight(PortalCreateEvent event) {
+
+        Material clickedBlock = event.getClickedBlock().getBlockData().getMaterial();
+        Action action = event.getAction();
+        User user = ess.getUser(event.getPlayer());
```
Why use `User`? `PlayerInteractEvent#getItem` returns the stack that the player used.

werogg committed:
Convention fix

Co-Authored-By: md678685 <1917406+md678685@users.noreply.github.com>

werogg committed:
Config.yml comments fix

Co-Authored-By: md678685 <1917406+md678685@users.noreply.github.com>

werogg committed:
PlayerInteractEvent#getItem instead of using User Object

werogg committed:
Merge remote-tracking branch 'origin/patch-2' into patch-2

werogg committed:
v1_14_DYES usages fixed to V1_14_DYES

werogg committed:
Merge branch 'patch-2' of https://github.com/werogg/Essentials into patch-2

pop4959 commented:
Is this ready? Title is kind of botched, btw.

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/utils/MaterialUtil.java:
```diff
@@ -86,6 +96,11 @@ public static boolean isLeatherArmor(Material material) {
     }
 
+    public static boolean isDye(Material material) {
+        return VersionUtil.getServerBukkitVersion().isHigherThanOrEqualTo(VersionUtil.v1_14_R01) ?
+                CLASSIC_DYES.contains(material) : CLASSIC_DYES.contains(material) || V1_14_DYES.contains(material);
```
```suggestion
        return CLASSIC_DYES.contains(material) || VersionUtil.getServerBukkitVersion().isHigherThanOrEqualTo(VersionUtil.v1_14_R01) ?
                 V1_14_DYES.contains(material): false;
```

Inversion!

pop4959 commented:
Closing this since the author appears to be inactive
